config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java

```diff
@@ -506,6 +514,56 @@ public void requestWhenSessionManagementConfiguredThenUserConfigurationOverrides
 		assertThat(result.getRequest().getSession(false)).isNotNull();
 	}
 
+	@Test
+	public void requestWhenCustomJwtDecoderWiredOnDslThenUsed()
+			throws Exception {
+
+		this.spring.register(CustomJwtDecoderOnDsl.class, BasicController.class).autowire();
+
+		this.mvc.perform(get("/authenticated")
+				.with(bearerToken("token")))
+				.andExpect(status().isOk())
+				.andExpect(content().string("custom-test-subject"));
+	}
+
+	@Test
+	public void requestWhenCustomJwtDecoderExposedAsBeanThenUsed()
+			throws Exception {
+
+		this.spring.register(CustomJwtDecoderAsBean.class, BasicController.class).autowire();
+
+		this.mvc.perform(get("/authenticated")
+				.with(bearerToken("token")))
+				.andExpect(status().isOk())
+				.andExpect(content().string("custom-test-subject"));
+	}
+
+	@Test
+	public void requestWhenConflictingJwtDecodersThenTheDslWiredOneTakesPrecedence()
+			throws Exception {
+
+		this.spring.register(CustomJwtDecoderConflicting.class, BasicController.class).autowire();
+
+		this.mvc.perform(get("/authenticated")
+				.with(bearerToken("token")))
+				.andExpect(status().isOk())
+				.andExpect(content().string("custom-test-subject"));
+	}
+
+	@Test
+	public void requestWhenJwtDecoderAndJwkSetUriConfiguredThenJwkSetUriTakesPrecedence()
```

Comment:
I'm wondering if this makes sense? re: `jwkSetUri` takes precedence over `JwtDecoder`. If the user supplied both of these configurations than we should at least log a warning message or something alerting the user. The user may wonder why their custom `JwtDecoder` is not wired?

Reply:
This is always the case with Spring Security. If you provide an explicit configuration via the DSL the bean is ignored. This makes sense since the bean is a default value.

Reply:
I asked @rwinch a similar question a while back about conflicting configurations (different scenario, though), so I'd like to hear his thoughts here.

At least, I don't think we should error, but I can see value in a logs that explain why we made this or that wiring decision. I know for my part, the "--debug" setting in Spring Boot has often been helpful to me when debugging why a given @Conditional-wiring passed or failed.

Reply:
@rwinch The test (and user) configures `jwkSetUri` via the DSL and registers a `JwtDecoder` `@Bean`. As a user, if my `JwtDecoder` `@Bean` is not leveraged than I'm left wondering why? This scenario should at least alert the user up-front about the configuration?

Reply:
I disagree. Should we warn a user if they invoke the same method twice? Should we warn a user if the configure a dsl and then disable it afterwards? This leads to too much information.

This style of configuration has worked very successfully for Spring Security.

Reply:
I think a message like this would be valuable at a debug level, similar to what Spring Boot does with the --debug flag, indicating why a given wiring decision was made.

That said, if we don't already have an established pattern for alerting folks to wiring choices (or that our pattern is that we don't alert them), then an isolated message like this would kind of come from nowhere. Probably better to consider the value of the pattern as a whole first (and at a later time).